controller/users.go

```diff
@@ -291,11 +270,178 @@ func (c *UsersController) Update(ctx *app.UpdateUsersContext) error {
+
+	updatedEmail := ctx.Payload.Data.Attributes.Email
+	if updatedEmail != nil && *updatedEmail != user.Email {
+		isUnique, err := isEmailUnique(appl, *updatedEmail, *user)
+		if err != nil {
+			return false, jsonapi.JSONErrorResponse(ctx, errs.Wrap(err, fmt.Sprintf("error updating idenitity with id %s and user with id %s", identity.ID, identity.UserID.UUID)))
+		}
+		if !isUnique {
+			jerrors, _ := jsonapi.ErrorToJSONAPIErrors(goa.ErrInvalidRequest(fmt.Sprintf("email address: %s is already in use", *updatedEmail)))
+			return false, ctx.Conflict(jerrors)
+		}
+		user.Email = *updatedEmail
+		err = c.populateExistingKeycloakProfileData(ctx, keycloakUserProfile)
+		if err != nil {
+			return false, err
+		}
+		keycloakUserProfile.Email = updatedEmail
+		return true, nil
+	}
+	return false, nil
+}
+
+func (c *UsersController) prepareToUpdateUsername(appl application.Application, ctx *app.UpdateUsersContext, u *account.User, identity *account.Identity, keycloakUserProfile *login.KeycloakUserProfile) (bool, error) {
+
+	updatedUserName := ctx.Payload.Data.Attributes.Username
+	if updatedUserName != nil && *updatedUserName != identity.Username {
+		if identity.RegistrationCompleted {
+			jerrors, _ := jsonapi.ErrorToJSONAPIErrors(goa.ErrInvalidRequest(fmt.Sprintf("username cannot be updated more than once for identity id %s ", identity.ID)))
+			return false, ctx.Forbidden(jerrors)
+		}
+		isUnique, err := isUsernameUnique(appl, *updatedUserName, *identity)
+		if err != nil {
+			return false, jsonapi.JSONErrorResponse(ctx, errs.Wrap(err, fmt.Sprintf("error updating idenitity with id %s and user with id %s", identity.ID, identity.UserID.UUID)))
+		}
+		if !isUnique {
+			jerrors, _ := jsonapi.ErrorToJSONAPIErrors(goa.ErrInvalidRequest(fmt.Sprintf("username : %s is already in use", *updatedUserName)))
+			return false, ctx.Conflict(jerrors)
+		}
+		identity.Username = *updatedUserName
+		identity.RegistrationCompleted = true
+		err = c.populateExistingKeycloakProfileData(ctx, keycloakUserProfile)
+		if err != nil {
+			return false, err
+		}
+		keycloakUserProfile.Username = updatedUserName
+		return true, nil
+	}
+	return false, nil
+}
+
+func (c *UsersController) prepareToUpdateRegistrationCompleted(appl application.Application, ctx *app.UpdateUsersContext, u *account.User, identity *account.Identity, keycloakUserProfile *login.KeycloakUserProfile) (bool, error) {
+
+	updatedRegistratedCompleted := ctx.Payload.Data.Attributes.RegistrationCompleted
+	if updatedRegistratedCompleted != nil && *updatedRegistratedCompleted != identity.RegistrationCompleted {
+		if !*updatedRegistratedCompleted {
+			jerrors, _ := jsonapi.ErrorToJSONAPIErrors(goa.ErrInvalidRequest(fmt.Sprintf("invalid value assigned to registration_completed for identity with id %s and user with id %s", identity.ID, identity.UserID.UUID)))
```

Comment:
The error message could be more clear saying registration completed attribute can't be set to false once it's set to true.